src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -199,11 +202,47 @@ private function parseDefinition(\DOMElement $service, $file, array $defaults =
             } elseif (isset($defaults['public'])) {
                 $public = $defaults['public'];
             }
-            $this->container->setAlias((string) $service->getAttribute('id'), new Alias($alias, $public));
+            $this->container->setAlias($id, new Alias($alias, $public));
 
             return;
         }
 
+        $definition = $this->getDefinition($service, $file, $defaults);
+        if (null === $instanceof) {
+            return $definition;
+        }
+
+        $className = $definition->getClass() ?: $id;
```

Comment:
We have 2 options here:

1. Don't support parameters as id when using _instanceof and document this behavior
2. Store the something like the following and the container and create child definitions in a compiler pass

```php
[
    '/foo/bar.yaml' => [
        'services' => ['id1', 'id2'],
        'instanceof' => ['type1' => new Definition(/*...*/), 'type2' => new Definition(/*...*/)]
    ]
```

For the sake of simplicity, I'm for the solution 1. Solution 2 will create new edges cases (when using inline definitions for instance).
Or maybe someone have a better idea?